crates/example-types/src/block_types.rs

```diff
@@ -17,38 +17,77 @@ use hotshot_types::{
 };
 use serde::{Deserialize, Serialize};
 use sha3::{Digest, Keccak256};
+use snafu::Snafu;
 use time::OffsetDateTime;
 
 use crate::{node_types::TestTypes, state_types::TestInstanceState};
 
 /// The transaction in a [`TestBlockPayload`].
 #[derive(Default, PartialEq, Eq, Hash, Serialize, Deserialize, Clone, Debug)]
-pub struct TestTransaction(pub Vec<u8>);
+#[serde(try_from = "Vec<u8>")]
+pub struct TestTransaction(Vec<u8>);
+
+#[derive(Debug, Snafu)]
+pub struct TransactionTooLong;
+
+impl TryFrom<Vec<u8>> for TestTransaction {
+    type Error = TransactionTooLong;
+
+    fn try_from(value: Vec<u8>) -> Result<Self, Self::Error> {
+        Self::try_new(value).ok_or(TransactionTooLong)
+    }
+}
 
 impl TestTransaction {
+    /// Construct new transaction
+    ///
+    /// # Panics
+    /// If `bytes.len()` > `u32::MAX`
+    pub fn new(bytes: Vec<u8>) -> Self {
```

Comment:
Would this be better served as `impl From<Vec<u8>>` or something? That way we could go from bytes to tx and back? Open to your thoughts.

Reply:
It already implements `TryFrom<Vec<u8>>`. Do you mean ditching constructor entirely in favor of that implementation?

Reply:
Yeah that's what I was thinking. But if you prefer to have both, I'm not opposed. Just trying to think of it from a rust-y standpoint.

Reply:
I think still having a constructor would be more rust-y, see for example all the wrapper types (`Cell`, `Arc`, `Mutex`, etc) in stdlib: they have corresponding `From` implementations, but provide `::new` as well.
Also, constructor is something that you almost definitely will want to call on a type, and to call `TestTransaction::try_from(x)` you'd have import `TryFrom` trait, which is a hurdle.